model/map-ldap/src/main/java/org/keycloak/models/map/storage/ldap/role/LdapRoleMapKeycloakTransaction.java

```diff
@@ -0,0 +1,441 @@
+        this.session = session;
+        this.config = config;
+        this.roleMapperConfig = new LdapMapRoleMapperConfig(config);
+        this.ldapMapConfig = new LdapMapConfig(config);
+    }
+
+    // interface matching the constructor of this class
+    public interface LdapRoleMapKeycloakTransactionFunction<A, B, R> {
+        R apply(A a, B b);
+    }
+
+    // TODO: entries might get stale if a DN of an entry changes due to changes in the entity in the same transaction
+    private final Map<String, String> dns = new HashMap<>();
+
+    public String readIdByDn(String dn) {
+        // TODO: this might not be necessary if the LDAP server would support an extended OID
+        // https://ldapwiki.com/wiki/LDAP_SERVER_EXTENDED_DN_OID
+
+        String id = dns.get(dn);
+        if (id == null) {
+            for (Map.Entry<String, LdapMapRoleEntityFieldDelegate> entry : entities.entrySet()) {
+                LdapMapObject ldap = entry.getValue().getLdapMapObject();
+                if (ldap.getDn().toString().equals(dn)) {
+                    id = ldap.getId();
+                    break;
+                }
+            }
+        }
+        if (id != null) {
+            return id;
+        }
+
+        LdapMapQuery ldapQuery = new LdapMapQuery();
+
+        // For now, use same search scope, which is configured "globally" and used for user's search.
+        ldapQuery.setSearchScope(ldapMapConfig.getSearchScope());
+        ldapQuery.setSearchDn(config.get("roles.dn.common"));
+
+        // TODO: read them properly to be able to store them in the transaction so they are cached?!
+        Collection<String> roleObjectClasses = ldapMapConfig.getRoleObjectClasses();
+        ldapQuery.addObjectClasses(roleObjectClasses);
+
+        String rolesRdnAttr = roleMapperConfig.getRoleNameLdapAttribute();
+
+        ldapQuery.addReturningLdapAttribute(rolesRdnAttr);
+
+        LdapMapIdentityStore identityStore = new LdapMapIdentityStore(session, ldapMapConfig);
+
+        LdapMapDn.RDN rdn = LdapMapDn.fromString(dn).getFirstRdn();
+        String key = rdn.getAllKeys().get(0);
+        String value = rdn.getAttrValue(key);
+
+        LdapRoleModelCriteriaBuilder mcb =
+                new LdapRoleModelCriteriaBuilder(roleMapperConfig).compare(RoleModel.SearchableFields.NAME, ModelCriteriaBuilder.Operator.EQ, value);
+        mcb = mcb.withCustomFilter(roleMapperConfig.getCustomLdapFilter());
+        ldapQuery.setModelCriteriaBuilder(mcb);
+
+        List<LdapMapObject> ldapObjects = identityStore.fetchQueryResults(ldapQuery);
+        if (ldapObjects.size() == 1) {
+            dns.put(dn, ldapObjects.get(0).getId());
+            return ldapObjects.get(0).getId();
+        }
+        return null;
+    }
+
+    private MapModelCriteriaBuilder<String, MapRoleEntity, RoleModel> createCriteriaBuilderMap() {
+        return new MapModelCriteriaBuilder<>(keyConverter, MapFieldPredicates.getPredicates(RoleModel.class));
```

Comment:
This method can override the `compare` method to account for `null` values in `REALM_ID` in case of `EQ` operator.

Reply:
@hmlnarik - that's an interesting idea. In order for that to work I'd need to pass in an instantiator to create new instances, as currently it creates only instances of MapModelCriteriaBuilder() in the tree created. All instances in the tree would need to be of the new type to ensure that they all handle the EQ operator differently.